internal/terraform/node_resource_plan.go

```diff
@@ -150,81 +140,77 @@ func (n *nodeExpandPlannableResource) DynamicExpand(ctx EvalContext) (*Graph, er
 		}
 	}
 
-	return &g, nil
-}
-
-// NodePlannableResource represents a resource that is "plannable":
-// it is ready to be planned in order to create a diff.
-type NodePlannableResource struct {
-	*NodeAbstractResource
-
-	Addr addrs.AbsResource
-
-	// ForceCreateBeforeDestroy might be set via our GraphNodeDestroyerCBD
-	// during graph construction, if dependencies require us to force this
-	// on regardless of what the configuration says.
-	ForceCreateBeforeDestroy *bool
-
-	// skipRefresh indicates that we should skip refreshing individual instances
-	skipRefresh bool
-
-	// skipPlanChanges indicates we should skip trying to plan change actions
-	// for any instances.
-	skipPlanChanges bool
-
-	// forceReplace are resource instance addresses where the user wants to
-	// force generating a replace action. This set isn't pre-filtered, so
-	// it might contain addresses that have nothing to do with the resource
-	// that this node represents, which the node itself must therefore ignore.
-	forceReplace []addrs.AbsResourceInstance
-
-	dependencies []addrs.ConfigResource
-}
-
-var (
-	_ GraphNodeModuleInstance       = (*NodePlannableResource)(nil)
-	_ GraphNodeDestroyerCBD         = (*NodePlannableResource)(nil)
-	_ GraphNodeDynamicExpandable    = (*NodePlannableResource)(nil)
-	_ GraphNodeReferenceable        = (*NodePlannableResource)(nil)
-	_ GraphNodeReferencer           = (*NodePlannableResource)(nil)
-	_ GraphNodeConfigResource       = (*NodePlannableResource)(nil)
-	_ GraphNodeAttachResourceConfig = (*NodePlannableResource)(nil)
-)
+	// The above dealt with the expansion of the containing module, so now
+	// we need to deal with the expansion of the resource itself across all
+	// instances of the module.
+	//
+	// We'll gather up all of the leaf instances we learn about along the way
+	// so that we can inform the checks subsystem of which instances it should
+	// be expecting check results for, below.
+	var diags tfdiags.Diagnostics
+	instAddrs := addrs.MakeSet[addrs.Checkable]()
+	for _, module := range moduleInstances {
+		resAddr := n.Addr.Resource.Absolute(module)
+		err := n.expandResourceInstances(ctx, resAddr, &g, instAddrs)
+		diags = diags.Append(err)
+	}
+	if diags.HasErrors() {
+		return nil, diags.ErrWithWarnings()
+	}
 
-func (n *NodePlannableResource) Path() addrs.ModuleInstance {
-	return n.Addr.Module
-}
+	// If this is a resource that participates in custom condition checks
+	// (i.e. it has preconditions or postconditions) then the check state
+	// wants to know the addresses of the checkable objects so that it can
+	// treat them as unknown status if we encounter an error before actually
+	// visiting the checks.
+	if checkState := ctx.Checks(); checkState.ConfigHasChecks(n.NodeAbstractResource.Addr) {
+		checkState.ReportCheckableObjects(n.NodeAbstractResource.Addr, instAddrs)
+	}
 
-func (n *NodePlannableResource) Name() string {
-	return n.Addr.String()
+	return &g, diags.ErrWithWarnings()
```

Comment:
Because the process here combines multiple distinct graphs, the graph here often won't have a single root. The single root probably doesn't matter in practice, that is vestigial from very early Terraform, but it does show that we're not actually validating the graph return from `DynamicExpand`. These have always been very simple and had no way to introduce cycles, but maybe since you're already touching the DynamicExpand call site it might be worth mentioning.

Reply:
Hmm indeed... I typically think of these `DynamicExpand` graphs as just an over-engineered way to iterate over the multiple expanded items, but you're right that the surrounding infrastructure at least _thinks_ it's dealing with single-rooted acyclic graphs even though these graphs never contain any real dependency edges in practice.

I'm going to first see if there's a straightforward way to add an additional root node just to maintain the contract, but if that isn't easy and low-risk then I'll find a suitable place to add a comment describing this hazard.

Reply:
Yeah, these have definitely been reduced to an obfuscated iteration, and in other cases we've removed the graph builder entirely. It does still have to get fed into the graph walker for concurrent operations, but maybe in the future we get rid of these last few specialized DynamicExpand graph transformers.

Reply:
While investigating this I noticed two interesting things:
* Because the test case I added here has only one instance of one resource in the module, the root transformer didn't add an extra root node anyway because that single resource instance was already the root.
* After adding another resource to force the addition of a root, I noticed that we do actually still end up with only one root because all of the root nodes are equal to one another as currently implemented: the node type is an empty struct and we use a value of that type (rather than a pointer) as the node. Therefore it naturally coalesces together into a single node anyway.

With those two observations in mind, I'm going to add a comment to the RootTransformer that records that all root node values being equal to each other is part of its contract and leave it at that for now. If we want to do something more involved later on then of course we can, but the current code is generating a valid graph already and I don't want to add more complexity to it just for a bug fix.

Reply:
If there are multiple levels of expansion, you still end up with multiple roots (I verified by calling `Validate` on the expanded subgraphs). I'm not saying this is necessarily something which needs to be fixed here, but we definitely don't always get a single root.

Reply:
Ahh, I see what the problem is: it's actually related to the _first_ observation I made above.

If there is only one resource instance in a particular module instance then `RootTransformer` won't insert a new root node at all, because there's already only one node in the graph. But that node isn't a root node so when we subsume the graphs all together we end up with just a pile of resource instances and no root, which therefore fails validation.

I managed to defeat my own reproduction by already having changed the test to generate multiple resource instances per resource before I tried calling `.Validate` on the result, and therefore masked the problem. :facepalm: 

It does seem to work regardless and so I'm going to go ahead with this change as a minimal v1.3 backport but I'll open a follow-up PR (intended only for the `main` branch, not backporting) to actually validate the `DynamicExpand` results and then make sure it does indeed produce valid graphs in all cases.